Entity/steering_behaviour.cpp

```diff
@@ -0,0 +1,221 @@
+//------------------------------------------------------------------------
+glm::vec3 SteeringBehaviour::Seek(glm::vec3 targetPosition) {
+    glm::vec3 desiredVelocity = glm::normalize(targetPosition - m_pEntity->GetPosition()) * m_pEntity->GetMaxSpeed();
+
+    return (desiredVelocity - m_pEntity->GetVelocity());
+}
+
+//----------------------------- Flee -------------------------------------
+//
+//  Does the opposite of Seek
+//------------------------------------------------------------------------
+glm::vec3 SteeringBehaviour::Flee(glm::vec3 targetPosition) {
+    //only flee if the target is within 'panic distance'. Work in distance
+    //squared space.
+    /* const double PanicDistanceSq = 100.0f * 100.0;
+  if (Vec2DDistanceSq(m_pEntity->GetPosition(), target) > PanicDistanceSq)
+  {
+    return glm::vec3(0,0);
+  }
+  */
+
+    glm::vec3 desiredVelocity = glm::normalize(m_pEntity->GetPosition() - targetPosition) * m_pEntity->GetMaxSpeed();
+
+    return (desiredVelocity - m_pEntity->GetVelocity());
+}
+
+//--------------------------- Arrive -------------------------------------
+//
+//  This behavior is similar to seek but it attempts to arrive at the
+//  target with a zero velocity
+//------------------------------------------------------------------------
+glm::vec3 SteeringBehaviour::Arrive(glm::vec3 targetPosition, Deceleration deceleration) {
+    glm::vec3 toTarget = targetPosition - m_pEntity->GetPosition();
+
+    //calculate the distance to the target
+    double dist = glm::length(toTarget);
+
+    if ( dist > 0 ) {
+        //because Deceleration is enumerated as an int, this value is required
+        //to provide fine tweaking of the deceleration..
+        const double decelerationTweaker = 1000.0f;
+
+        //calculate the speed required to reach the target given the desired deceleration
+        double speed = dist / ((double)deceleration * decelerationTweaker);
+
+        //make sure the velocity does not exceed the max
+        speed = glm::min(speed, (double)m_pEntity->GetMaxSpeed());
+
+        //from here proceed just like Seek except we don't need to normalize
+        //the toTarget vector because we have already gone to the trouble
+        //of calculating its length: dist.
+        glm::vec3 desiredVelocity = toTarget * (float)speed / (float)dist;
+
+        return (desiredVelocity - m_pEntity->GetVelocity());
+    }
+
+    return glm::vec3(0.0f, 0.0f, 0.0f);
+}
+glm::vec3 GetNormalPoint(glm::vec3 p, glm::vec3 a, glm::vec3 b) {
+    glm::vec3 ap = p - a;
+    glm::vec3 ab = b - a;
+    ab = glm::normalize(ab);
+    float dot = glm::dot(ap, ab);
+    ab = ab * dot;
+    return a + ab;
+}
+
+bool InSegmentRange(glm::vec3 start, glm::vec3 end, glm::vec3 point) {
+    float segmentLength = glm::distance(start, end);
+    float startToPoint = glm::distance(start, point);
+    float endToPoint = glm::distance(end, point);
+    return startToPoint + endToPoint <= segmentLength;
+}
```

Comment:
i wonder where we should move these. probably also into the math helpers right?

Reply:
Your decision. But I think that would make sense!